google/cloud/bigtable/internal/instance_admin.h

```diff
@@ -114,6 +114,54 @@ class InstanceAdmin {
   std::vector<google::bigtable::admin::v2::Instance> ListInstances(
       grpc::Status& status);
 
+  /**
+   * Makes an asynchronous request to get the list of instances.
+   *
+   * @param cq the completion queue that will execute the asynchronous calls,
+   *     the application must ensure that one or more threads are blocked on
+   *     `cq.Run()`.
+   * @param callback a functor to be called when the operation completes. It
+   *     must satisfy (using C++17 types):
+   *     static_assert(std::is_invocable_v<
+   *         Functor, google::bigtable::admin::v2::ListInstancesResponse&,
+   *         grpc::Status const&>);
+   *
+   * @tparam Functor the type of the callback.
+   */
+  template <typename Functor,
+            typename std::enable_if<
+                google::cloud::internal::is_invocable<
+                    Functor, CompletionQueue&,
+                    google::bigtable::admin::v2::ListInstancesResponse&,
+                    grpc::Status&>::value,
+                int>::type valid_callback_type = 0>
+  void AsyncListInstances(CompletionQueue& cq, Functor&& callback) {
```

Comment:
This is just making one call to fetch the instances, if you look at the blocking version:

https://github.com/GoogleCloudPlatform/google-cloud-cpp/blob/8a54c7520df824e1fcd7b73c4da7916f6be69820/google/cloud/bigtable/internal/instance_admin.cc#L32

they make multiple calls in a loop.  You need to unroll that loop into a class that keeps the application callback, and the accumulated state for all the responses, and then invokes the callback with the final accumulated result.

If you read #1262 I requested that you propose an API first, I think this should be:

```C++
  template <typename Functor,
            typename std::enable_if<
                google::cloud::internal::is_invocable<
                    Functor, CompletionQueue&,
                    std::vector<google::bigtable::admin::v2::Instance>&,
                    grpc::Status&>::value,
                int>::type valid_callback_type = 0>
  void AsyncListInstances(CompletionQueue& cq, Functor&& callback) {
```